SI-7974 Fix broken 'Symbol-handling code in CleanUp

soc opened the pull request:
The transformation did never happen because the pattern
failed to match. Why did the pattern fail to match? Because
Symbol_apply turned into an overloaded Symbol after erasure.

soc committed:
SI-7974 Clean up and test 'Symbol-handling code in CleanUp

Looks like the transformation did never happen because the pattern
failed to match. Why did the pattern fail to match? Because the
Symbol.apply we see in the tree claims to be a method while
Symbol_apply defined in Definitions wants to be a value.

This issue was caused because nonPrivateMember starts spitting out
overloaded symbols after erasure.
This issue has been fixed in the earlier commit, so what happens in
this commit is adding tests and fixing documentation.

soc commented:
@retronym 
- [x] determine if the original code ever worked. If not, when did it start failing? >> It worked in 2.8.2 and 2.9.3, it doesn't work anymore in 2.10.3.
- [ ] I think it is broken for {???; Symbol}.apply("foo") ... >> doesn't the use of `fullName` prevent that?
- [ ] needs to be discussed ... >> I'll ask on the mailing list.
- [ ] deeper understanding of the two symbols for apply you encounter. ... >> The culprit seems to be https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/Definitions.scala#L1223

soc commented:
Maybe @xeno-by has an idea about https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/Definitions.scala#L1223 ... git blame says that the comments are from him.

xeno-by commented:
Yeah, this behavior has been puzzling me as well, but I haven't been able
to find out why it happens.

On 18 November 2013 02:23, soc notifications@github.com wrote:

> Maybe @xeno-by https://github.com/xeno-by has an idea about
> https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/Definitions.scala#L1223... git blame says that the comments are from him.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3149#issuecomment-28669976
> .

retronym commented:
`Symbol_apply` calls `SymbolModule.info.nonPrivateMember(TermName("apply"))`, which returns an overloaded symbol containing the alternatives `Symbol#apply` and `UniquenessCache#apply`.

```
scala> :power

scala> enteringErasure(getMemberMethod(SymbolModule, nme.apply).alternatives.map(SymbolModule.info memberType _))
res22: List[$r.intp.global.Type] = List((name: String)Symbol)

scala> exitingErasure(getMemberMethod(SymbolModule, nme.apply).alternatives.map(SymbolModule.info memberType _))
res23: List[$r.intp.global.Type] = List((name: String)Symbol, (name: Object)Object)

scala> exitingErasure(Symbol_apply).alternatives.map(_.debugLocationString)
res9: List[String] = List(override method apply in object Symbol, method apply in class UniquenessCache)
```

It would seem that any call to `findMember` `exitingErasure` or later is a possible bug. You could wrap that call to the most localized fix for this PR is to wrap the call to `Symbol_apply` in `exitingTyper(...)`.

You could add a followup commit with a more wide-spread fix

Change:

``` scala
def getMemberIfDefined(owner: Symbol, name: Name): Symbol =
      owner.info.nonPrivateMember(name)
```

To:

``` scala
def getMemberIfDefined(owner: Symbol, name: Name): Symbol =
      // findMember considered harmful after erasure; e.g. 
      // 
      // scala> exitingErasure(Symbol_apply).isOverloaded
      // res27: Boolean = true
      //
      enteringPhaseNotLaterThan(currentRun.erasurePhase)(
        owner.info.nonPrivateMember(name)
      )
```

And remove the localised fix, and @xeno-by's comment workaround in `getMember`

Back to the motivating topic of this PR (static field initializers not going into `<clinit>`), I notice that the caches for structural types are also static fields, and do indeed appear to be emitted in the right way:

```
% cat sandbox/symbol.scala && qbin/scalac sandbox/symbol.scala && javap -p -v C
class C {
  def foo = 0

  def bar: Any = (this: { def foo: Any}).foo
}
warning: there were 1 feature warning(s); re-run with -feature for details
one warning found

Compiled from "symbol.scala"
public class C extends java.lang.Object
...
{
private static java.lang.Class[] reflParams$Cache1;

private static volatile java.lang.ref.SoftReference reflPoly$Cache1;

public static {};
  Code:
   Stack=4, Locals=0, Args_size=0
   0:   iconst_0
   1:   anewarray   #12; //class java/lang/Class
   4:   putstatic   #16; //Field reflParams$Cache1:[Ljava/lang/Class;
   7:   new #18; //class java/lang/ref/SoftReference
   10:  dup
   11:  new #20; //class scala/runtime/EmptyMethodCache
   14:  dup
   15:  invokespecial   #23; //Method scala/runtime/EmptyMethodCache."<init>":()V
   18:  invokespecial   #26; //Method java/lang/ref/SoftReference."<init>":(Ljava/lang/Object;)V
   21:  putstatic   #30; //Field reflPoly$Cache1:Ljava/lang/ref/SoftReference;
   24:  return
  LineNumberTable:
   line 4: 0

...

}

```

soc commented:
@retronym Wow, thanks a lot for debugging this! Just trying to understand this correctly ... why do we gain that second member after erasure in the first place?

soc commented:
> Back to the motivating topic of this PR (static field initializers not going into <clinit>), I notice that the caches for structural types are also static fields, and do indeed appear to be emitted in the right way: ...

Yes, I think the pattern is that we have a few specialized locations were work is done to make statics work correctly, but we are missing the bigger picture. I'm already looking into https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/transform/Constructors.scala#L634 were I think the appropriate general check is missing.

soc commented:
@retronym 

> enteringPhaseNotLaterThan(currentRun.erasurePhase)(...

Won't this cause problems because `currentRun` lives on `global` and `global` is in the compiler jar, not the reflection jar?

Is something like https://github.com/scala/scala/blob/master/src/reflect/scala/reflect/internal/util/TraceSymbolActivity.scala#L90 a better way?

retronym commented:
You are right. We can still get to it with the method used in `src/reflect/scala/reflect/internal/util/TraceSymbolActivity.scala`; we should refactor the private method `runBeforeErasure` out to be a member of `SymbolTable`.

soc committed:
SI-7974 Avoid calling nonPrivateMember after erasure

One example were this would lead to subtle bugs otherwise is
Symbol_apply, where after erasure an overloaded symbol containing
Symbol#apply and UniquenessCache#apply is returned.

    // findMember considered harmful after erasure; e.g.
    //
    // scala> exitingErasure(Symbol_apply).isOverloaded
    // res27: Boolean = true

soc commented:
Looks like 144e416 is failing due to whitespace differences ... is there some magic incantation for that, or should I just call `split(" ")` on everything?

soc commented:
_sigh_ ... JavapTest is pretty terrible to debug ... I think I finally figured it out.

som-snytt commented:
I was going to say something like, Wow, someone is actually using JavapTest, but I didn't want to discourage you.

Was it the `bejahen` thing that grabbed you?

soc commented:
No, it just feels like the implementation does extra work in case of an failure to not actually tell anyone what the problem was ... but I figured it out now. :-)

retronym commented:
You should actually use a bytecodetest to decouple from the rep.

On Wednesday, November 20, 2013, soc wrote:

> No, it just feels like the implementation does extra work in case of an
> failure to not actually tell anyone what the problem was ... but I figured
> it out now. :-)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3149#issuecomment-28861963
> .

adriaanm commented:
PLS REBUILD ALL

https://github.com/typesafehub/sbt-builds-for-ide/commit/08cfc9d1c40ac146be4373452e013e083eb463ef should fix ide build breakage (in my local testing, sbt 0.13 was the default)

scala-jenkins commented:
(kitty-note-to-self: ignore 28868185)
:cat: Roger! Rebuilding pr-scala for 1b804786, 37109b4f. :rotating_light:

soc commented:
PLS REBUILD/pr-scala@1b80478

scala-jenkins commented:
(kitty-note-to-self: ignore 28986460)
:cat: Roger! Rebuilding pr-scala for 1b804786. :rotating_light:

soc commented:
PLS REBUILD/pr-scala@37109b4

soc commented:
If this fails because javap decided to emit ”extends java.lang.Object” in javap 6 but stopped doing that in javap 7, I'll pull my hair out ...

scala-jenkins commented:
(kitty-note-to-self: ignore 28990076)
:cat: Roger! Rebuilding pr-scala for 37109b4f. :rotating_light:

soc commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 29077590)
:cat: Roger! Rebuilding pr-scala for 1b804786, 3f05d2f8. :rotating_light:

soc commented:
Wow, this must have been my most painful, nerve-wrecking test case ever.
@retronym Is this good to go?

adriaanm commented:
Thanks for persevering, @soc!

retronym commented:
We should not involve the REPL and `:javap` in this test case. It should be written as `ByteCodeTest`. Sorry my earlier comment didn't make the details clearer, I sent it from my phone.

Here's an example that used ASM to output the decompiled bytecode and compares it against a checkfile:

  https://github.com/retronym/scala/commit/178c37bdded#diff-9821ea5ae9849e5bd4c52285ec2e56f5R5

retronym commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -1208,15 +1208,21 @@ trait Definitions extends api.StandardDefinitions {
       }
     }
 
+    private lazy val erasurePhase = findPhaseWithName("erasure")
```
This _might_ be brittle in a compiler that doesn't have an "erasure" phase. (Maybe ScalaJS?)

How about?

``` scala
  final def findPhaseBeforeErasure: Phase = {
     var ph = phase
     while (ph != NoPhase && ph.erasedTypes) {
       ph = ph.prev
     }
     if (ph eq NoPhase) phase else ph
   }
  def enteringPhaseNotLaterThanErasure[T](op: => T): T = 
    enteringPhaseNotLaterThan(findPhaseBeforeErasure)(op)
  ...
  enteringPhaseNotLaterThanErasure(owner.info.nonPrivateMember(name))
```

soc replied:
Wouldn't this skip the phase exactly before erasure, too?

retronym replied:
Let's leave that determination up to a little test case. If you're right, I'll just say that it only ever intended as pseudocode :smiley:

soc replied:
Let's hope this is the last time I have to touch this code ... otherwise I'm pulling my hair out. :-)

retronym commented:
@soc let me know if you need a hand converting this REPL test to a BytecodeTest.

soc commented:
@retronym Replaced JavapTest with BytecodeTest. Is this better now?

retronym commented:
LGTM.

Sorry about the delay in reviewing.

/cc @axel22. I'm curious what sort of code you originally encountered that had bottlenecks from Symbol lookup that warranted this hoisting in the first place. Why not just use string literals in performance sensitive code?

retronym commented on test/files/run/t7974/Test.scala:
```diff
@@ -0,0 +1,20 @@
+    val result =
+      classString.split('\n')
+        .dropWhile(elem => elem != "public class Symbols {")
+        .filterNot(elem => elem.startsWith("  @Lscala/reflect/ScalaSignature") || elem.startsWith("  ATTRIBUTE ScalaSig"))
```
Nitpick: this test could have been a touch more focussed by doing this filtering at the ASM level, rather than with Strings.

Further more, the checkfile contains some irrelevant details, like line numbers (try `skipDebugInfo = true` next time).

But I'll give you a pass this time as this PR has been spinning for too long.
